Setup: able to download translation files automatically

jaapjansma opened the pull request:
Overview
----------------------------------------

With this PR the CiviCRM setup utility will download the translation files automatically. So making the download of the l10n civicrm files an obsolete step.

Before
----------------------------------------

One had to download the civicrm-l10n files before installing CiviCRM. Otherwise CiviCRM could not be installed into another language then English.

After
----------------------------------------

The setup utility now downloads the translation file for the selected language (if other than english).

Technical Details
----------------------------------------

The setup plugin creates a directory under civicrm-core/l10n

Comments
----------------------------------------

Thanks for the good documentation of the setup utility that made making this change pretty straight forward!

demeritcowboy commented on setup/plugins/installFiles/l10nDownload.civi-setup.php:
```diff
@@ -0,0 +1,26 @@
+      $civicrmMoFileUrl = 'https://download.civicrm.org/civicrm-l10n-core/mo/' . $lang . '/civicrm.mo';
+      $dir = $e->getModel()->srcPath . DIRECTORY_SEPARATOR . 'l10n' . DIRECTORY_SEPARATOR . $lang . DIRECTORY_SEPARATOR . 'LC_MESSAGES';
+      \Civi\Setup::log()->info("Creating directry: " . $dir);
+      \CRM_Utils_File::createDir($dir, FALSE);
```
If running via the web page, this might fail on sites with tighter security where the webserver user doesn't have write access to the code files tree. It should maybe log a warning and skip the remaining steps.

jaapjansma replied:
I have added error handling

demeritcowboy commented on setup/plugins/installFiles/l10nDownload.civi-setup.php:
```diff
@@ -0,0 +1,26 @@
+      \Civi\Setup::log()->info("Creating directry: " . $dir);
+      \CRM_Utils_File::createDir($dir, FALSE);
+      \Civi\Setup::log()->info("Download translation files from " . $civicrmMoFileUrl . ' into ' . $dir);
+      file_put_contents($dir . DIRECTORY_SEPARATOR . 'civicrm.mo', file_get_contents($civicrmMoFileUrl));
```
Similarly I don't know how common it is to allow urls in file_get_contents(). Some sites might fail here. Guzzle is usually used for downloads now.

jaapjansma replied:
I am now using Guzzle and have added some error detection.

demeritcowboy commented on setup/plugins/installFiles/l10nDownload.civi-setup.php:
```diff
@@ -0,0 +1,26 @@
+    if ($lang || $lang != 'en_US') {
+      $civicrmMoFileUrl = 'https://download.civicrm.org/civicrm-l10n-core/mo/' . $lang . '/civicrm.mo';
+      $dir = $e->getModel()->srcPath . DIRECTORY_SEPARATOR . 'l10n' . DIRECTORY_SEPARATOR . $lang . DIRECTORY_SEPARATOR . 'LC_MESSAGES';
+      \Civi\Setup::log()->info("Creating directry: " . $dir);
```
Typo: "directry".

jaapjansma replied:
Fixed typo

jaapjansma commented:
@demeritcowboy thanks for your review. I have updated the installer with things from your review. Also I added a setting to the model for the download path and the creation of the l10n directory in the private files path (I had to set this as well). 
The good thing is that the l10n files are now downloaded outside of civicrm-core and are kept when civicrm is updated. And those files can also be updated with the l10n update extension.

demeritcowboy commented:
I'll give it a run.
The L10N_BASEDIR reminds me there's a path `[civicrm.l10n]/` that maybe could/should be used here, but I've had problems with the define getting ignored before. Anyway, will take a look.

jaapjansma commented:
Yes I know about the `[civicrm.l10n]` path but that one is undefined or not usable in the setup utility.  I tried that first.

totten commented:
* Pretty cool. This sounds like it would make the installation process easier!
* Agree you can't reliably consume `[civicrm.l10n]` at this point. Good call on materializing `[civicrm.private]`. FWIW, it could optionally read `$model->paths['civicrm.l10n']`. (*If the caller requests a specific path, then use it -- that seems elegant but not critical.*)
* I think the signature for the array could be improved:

    ```php
    $model->moFiles['civicrm.mo'] = 'https://download.civicrm.org/civicrm-l10n-core/mo/[locale]/civicrm.mo',
    $model->moFiles['cdntaxreceipts.mo'] = 'https://download.civicrm.org/civicrm-l10n-extensions/mo/cdntaxreceipts/[locale]/cdntaxreceipts.mo',
    ```

    * This gives some added purpose to using the `array`.
    * Note that some sources may not provide `en_US` files/urls, as in the case of `cdntaxreceipts`. Also, there are several
    [precedents](https://github.com/civicrm/civicrm-core/blob/5.68/xml/schema/Activity/Activity.xml#L13-L16) for URL-variables with `[var]` notation.
    * I think names `moFiles` or `moUrls` or `l10nFiles` would match the style of other names (`cmsDb`, `cmsBaseUrl`, `paths`) more closely. Also avoids typo.
* There's a discrepancy about when it does work:
    * The file is placed in `plugins/installFiles/` which implies that the *general idea* is to do work during `civi.setup.installFiles`.
    * The call to `addListener()` only registers one listener -- for `civi.setup.checkRequirements`.
    * We could do it like `InstallSettingsFile.civi-setup.php` - which has two parts:
        * There is [a `checkRequirements` part](https://github.com/civicrm/civicrm-core/blob/5.68/setup/plugins/installFiles/InstallSettingsFile.civi-setup.php#L16)
          which passively validates.
        * There is [a `installFiles` part`](https://github.com/civicrm/civicrm-core/blob/5.68/setup/plugins/installFiles/InstallSettingsFile.civi-setup.php#L62)
          that does the main work.

I realize that's a lot of suggestions. You don't have to go and implement all of them -- but it would be good to classify them. (Compare: "Suggestions you agree with and plan to work on" vs "Suggestions you agree with but don't have time for" vs "Suggestions you disagree with because $reasons")

jaapjansma commented:
Thanks @totten for the feedback. I have implemented your suggestions. That made sense to me!

totten commented:
@jaapjansma That's pretty cool. I've been playing around with it on the CLI installer. My testing is like this:

* (Re)move my copy of `civicrm/l10n/` source tree (where I usually get data)
* Alternately reset the DB and run the CLI installer

    ```bash
    civibuild restore dmaster ; cv core:uninstall -f 
    cv core:install --lang=nl_NL --cms-base-url=http://dmaster.bknix:8001 -vv
    civibuild restore dmaster ; cv core:uninstall -f 
    cv core:install --lang=fr_FR --cms-base-url=http://dmaster.bknix:8001 -vv
    civibuild restore dmaster ; cv core:uninstall -f 
    cv core:install --lang=de_DE --cms-base-url=http://dmaster.bknix:8001 -vv
    ```
* And it seems to work as advertised. :smile: 

Flagging `merge ready` (in case @demeritcowboy or @mlutfy has more to add).

There might be a couple tweaks to consider separately -  e.g. maybe `AvailableLangs.civi-setup.php` should offer all langs to the GUI installer (since it can download them); e.g. maybe it should skip downloading there's an available that's relatively recent (to avoid unnecessary downloads on local dev-builds and Drupal multisite). None of that needs to block PR IMHO.

demeritcowboy commented on setup/plugins/installFiles/l10nDownload.civi-setup.php:
```diff
@@ -0,0 +1,69 @@
+      // Set l10n basedir as a define. The GenCode.php tries to locate the l10n files
+      // from this location if other than l10n in the civicrm core directory.
+      if (!isset($e->getModel()->paths['civicrm.l10n']['path'])) {
+        if (defined('CIVICRM_L10N_BASEDIR')) {
```
How would I `define` this? In drupal settings.php? Maybe use `CRM_Utils_Constant::value()` and then I can set it with an environment variable.

jaapjansma replied:
@demeritcowboy I had another look and changed it to use the `CRM_Utils_Constant`. I also changed setting the define to setting the environment variable.

jaapjansma committed:
added plugin to download l10n translation files

demeritcowboy commented:
This is pretty great.

I only tested drupal 7 and 10. Agree that the UI will need some tweaks but doesn't block the PR.

One thing that is confusing for both UI and CLI is that unless they know about this new feature or are using CIVICRM_L10N_BASEDIR the average person is going to do the normal thing and put the files ahead of time in sites/all/modules/civicrm/l10n, but then it puts another copy in sites/default/files/civicrm/l10n, and then when they go to update the translations in the future they'll do what they always do and update sites/all/modules/civicrm/l10n, but it will get ignored and it will still use the old translation in sites/default/files/civicrm/l10n. There may be much head-scratching.

peteyno36 commented:
Nice

fuentesvazquezjoseeduardo commented:
Y
